Scripts/download-portal-item-data.swift

```diff
@@ -100,49 +138,104 @@ struct PortalItem: Decodable {
+        exit(1)
+    }
+}()
+let destinationURLProvider = DestinationURLProvider(downloadDirectory: downloadDirectoryURL, fileTypes: fileTypes)
+
+typealias Identifier = String
+typealias Filename = String
+typealias DownloadedItems = [Identifier: Filename]
+
+let downloadedItemsURL = downloadDirectoryURL.appendingPathComponent(".downloaded_items", isDirectory: false)
+let previousDownloadedItems: DownloadedItems = {
+    do {
+        let data = try Data(contentsOf: downloadedItemsURL)
+        let decoder = PropertyListDecoder()
+        return try decoder.decode(DownloadedItems.self, from: data)
+    } catch {
+        return [:]
+    }
+}()
+var downloadedItems = previousDownloadedItems
+
 let dispatchGroup = DispatchGroup()
 
 portalItems.forEach { (portalURLString, portalItems) in
     let portalURL = URL(string: portalURLString)!
     portalItems.forEach { (portalItem) in
-        let destinationURL = makeDownloadURL(downloadDirectoryURL: downloadDirectoryURL, filename: portalItem.filename)
-        guard !FileManager.default.fileExists(at: destinationURL) else {
-            return
-        }
-        
-        dispatchGroup.enter()
-        
-        print("Downloading \(portalItem.filename)")
-        fflush(stdout)
-        let sourceURL = makeDataURL(portalURL: portalURL, itemIdentifier: portalItem.identifier)
-        downloadFile(at: sourceURL, to: destinationURL) { (error) in
-            if let error = error {
-                print("Error downloading \(portalItem.filename): \(error)")
-                URLSession.shared.invalidateAndCancel()
-                exit(1)
-            } else {
-                dispatchGroup.leave()
+        // Have we already downloaded the item?
+        let filename = downloadedItems[portalItem.identifier] ?? portalItem.filename
+        if FileManager.default.fileExists(atPath: destinationURLProvider.makeURL(filename: filename).path) {
+            print("Item \(portalItem.identifier) has already been downloaded")
+            // This is a temporary measure for users who currently don't have a
+            // downloaded items file.
+            downloadedItems[portalItem.identifier] = filename
+        } else {
+            print("Downloading item \(portalItem.identifier)")
+            fflush(stdout)
+            
+            dispatchGroup.enter()
+            let sourceURL = makeDataURL(portalURL: portalURL, itemIdentifier: portalItem.identifier)
+            downloadFile(at: sourceURL, destinationURLProvider: destinationURLProvider) { (result) in
+                switch result {
+                case .success(let url):
+                    // ' + 1' removes the leading path separator.
+                    downloadedItems[portalItem.identifier] = url.lastPathComponent
+                    dispatchGroup.leave()
+                case .failure(let error):
+                    print("warning: Error downloading item \(portalItem.identifier): \(error)")
```

Comment:
need a `dispatchGroup.leave()` here?

Reply:
Not necessary since we go on to call `exit(_:)`, which returns `Never`.